File: Javascript/13.roman-to-integer.js
/*
 * @lc app=leetcode id=13 lang=javascript
 *
 * [13] Roman to Integer
 */

// @lc code=start
/**
 * @param {string} s
 * @return {number}
 */
var romanToInt = function(s) {
    const map = new Map([["I", 1], ["V", 5], ["X", 10], ["L", 50], ["C", 100], ["D", 500], ["M", 1000]]);

    let num = 0;
    let i = 0;
    while (i < s.length) {
        const current = map.get(s.charAt(i));
        const next = map.get(s.charAt(i + 1));

        if (next > current) {
            num += ((next / current) - 1) * current;
            i += 2;
            continue;
        }

        num += current;
        i++;
        continue;
    }
    return num;
};
// @lc code=end

File: Ruby/13.roman-to-integer.rb
#
# @lc app=leetcode id=13 lang=ruby
#
# [13] Roman to Integer
#

# @lc code=start
# @param {String} s
# @return {Integer}
def roman_to_int(s)
    map = { "I" => 1, "V" => 5, "X" => 10, "L" => 50, "C" => 100, "D" => 500, "M" => 1000 }

    num = 0
    i = 0

    while i < s.size
        currentChar = map[s[i]]
        nextChar = map[s[i + 1]]

        if nextChar == nil
            nextChar = 0
        end

        if nextChar > currentChar
            num += ((nextChar / currentChar) - 1) * currentChar
            i += 2
            next
        end

        num += currentChar
        i += 1
    end

    return num
end
# @lc code=end

roman_to_int ""